tests/unit/completion/test_completer.py

```diff
@@ -0,0 +1,189 @@
+    }
+    instances[usertypes.Completion.value] = {
+        'general': {
+            'ignore-case': FakeCompletionModel(usertypes.Completion.value),
+        }
+    }
+    monkeypatch.setattr('qutebrowser.completion.completer.instances',
+                        instances)
+
+
+@pytest.fixture(autouse=True)
+def cmdutils_patch(monkeypatch, stubs):
+    """Patch the cmdutils module to provide fake commands."""
+    cmds = {
+        'set': [usertypes.Completion.section, usertypes.Completion.option,
+                usertypes.Completion.value],
+        'help': [usertypes.Completion.helptopic],
+        'quickmark-load': [usertypes.Completion.quickmark_by_name],
+        'bookmark-load': [usertypes.Completion.bookmark_by_url],
+        'open': [usertypes.Completion.url],
+        'buffer': [usertypes.Completion.tab],
+        'session-load': [usertypes.Completion.sessions],
+        'bind': [usertypes.Completion.empty, usertypes.Completion.command],
+        'tab-detach': None,
+    }
+    cmd_utils = stubs.FakeCmdUtils({
+        name: stubs.FakeCommand(completion=compl)
+        for name, compl in cmds.items()
+    })
+    monkeypatch.setattr('qutebrowser.completion.completer.cmdutils',
+                        cmd_utils)
+
+
+@pytest.mark.parametrize('txt, expected', [
+    (':nope|', usertypes.Completion.command),
+    (':nope |', None),
+    (':set |', usertypes.Completion.section),
+    (':set gen|', usertypes.Completion.section),
+    (':set general |', usertypes.Completion.option),
+    (':set what |', None),
+    (':set general ignore-case |', usertypes.Completion.value),
+    (':set general huh |', None),
+    (':help |', usertypes.Completion.helptopic),
+    (':quickmark-load |', usertypes.Completion.quickmark_by_name),
+    (':bookmark-load |', usertypes.Completion.bookmark_by_url),
+    (':open |', usertypes.Completion.url),
+    (':buffer |', usertypes.Completion.tab),
+    (':session-load |', usertypes.Completion.sessions),
+    (':bind |', usertypes.Completion.empty),
+    (':bind <c-x> |', usertypes.Completion.command),
+    (':bind <c-x> foo|', usertypes.Completion.command),
+    (':bind <c-x>| foo', usertypes.Completion.empty),
+    (':set| general ', usertypes.Completion.command),
+    (':|set general ', usertypes.Completion.command),
+    (':set gene|ral ignore-case', usertypes.Completion.section),
+    (':|', usertypes.Completion.command),
+    (':   |', usertypes.Completion.command),
+    (':bookmark-load      |', usertypes.Completion.bookmark_by_url),
+    ('/|', None),
+    (':open -t|', None),
+    (':open --tab|', None),
+    (':open -t |', usertypes.Completion.url),
+    (':open --tab |', usertypes.Completion.url),
+    (':open | -t', usertypes.Completion.url),
+    (':--foo --bar |', None),
+    (':tab-detach |', None),
+    (':bind --mode=caret <c-x> |', usertypes.Completion.command),
+    #(':bind --mode caret <c-x> |', usertypes.Completion.command), KNOWN BUG
```

Comment:
Ah, cool. Then it prevents you from forgetting to re-enable the test when its fixed. Good to know.